cdc/kv/shared_client.go

```diff
@@ -0,0 +1,1233 @@
+		zap.String("changefeed", s.changefeed.ID),
+		zap.Any("subscriptionID", subscriptionID),
+		zap.Uint64("ResolvedTs", resolvedTs.Ts),
+		zap.Int("regionCount", len(resolvedTs.Regions)))
+
+	for _, regionID := range resolvedTs.Regions {
+		slot := hashRegionID(regionID, len(s.workers))
+		if sfEvents[slot].stream == nil {
+			sfEvents[slot] = newResolvedTsBatch(resolvedTs.Ts, rs)
+		}
+		x := &sfEvents[slot].resolvedTsBatch
+		if state := rs.getState(subscriptionID, regionID); state != nil {
+			x.regions = append(x.regions, state)
+		}
+	}
+
+	for i, sfEvent := range sfEvents {
+		if len(sfEvent.resolvedTsBatch.regions) > 0 {
+			sfEvent.stream = rs
+			if err := s.workers[i].sendEvent(ctx, sfEvent); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+func (s *SharedClient) newStream(ctx context.Context, g *errgroup.Group, r *requestedStore) *requestedStream {
+	stream := &requestedStream{
+		streamID: streamIDGen.Add(1),
+		requests: chann.NewAutoDrainChann[singleRegionInfo](),
+	}
+
+	stream.requestedRegions.m = make(map[SubscriptionID]map[uint64]*regionFeedState)
+	stream.init = func(ctx context.Context) (err error) {
+		if stream.conn, err = s.grpcPool.GetConn(r.storeAddr); err != nil {
+			return errors.Trace(err)
+		}
+		if err = version.CheckStoreVersion(ctx, s.pd, r.storeID); err != nil {
+			return errors.Trace(err)
+		}
+		rpc := cdcpb.NewChangeDataClient(stream.conn.ClientConn)
+		ctx = getContextFromFeatures(ctx, []string{rpcMetaFeatureStreamMultiplexing})
+		if stream.client, err = rpc.EventFeedV2(ctx); err != nil {
+			return errors.Trace(err)
+		}
+		return nil
+	}
+
+	g.Go(func() error {
+		for {
+			select {
+			case <-ctx.Done():
+				return ctx.Err()
+			case sri := <-stream.requests.Out():
+				stream.preFetchForConnecting = new(singleRegionInfo)
+				*stream.preFetchForConnecting = sri
+			}
+
+			if canceled := s.runStream(ctx, r, stream); canceled {
+				return nil
+			}
+			for _, m := range stream.clearStates() {
+				for _, state := range m {
+					state.markStopped(&sendRequestToStoreErr{})
+					slot := hashRegionID(state.sri.verID.GetID(), len(s.workers))
+					sfEvent := statefulEvent{eventItem: eventItem{state: state}}
+					_ = s.workers[slot].sendEvent(ctx, sfEvent)
+				}
+			}
+			// Why we need to re-schedule pending regions? This because the store can
+			// fail forever, and all regions are scheduled to other stores.
+			for _, sri := range stream.clearPendingRegions() {
+				s.onRegionFail(ctx, newRegionErrorInfo(sri, &sendRequestToStoreErr{}))
+			}
+
+			if err := util.Hang(ctx, time.Second); err != nil {
+				// TODO(qupeng): handle the case that TiKV closes the connection.
+				return err
+			}
+		}
+	})
+
+	return stream
+}
+
+func (s *SharedClient) clearStream(r *requestedStore, stream *requestedStream) {
+	if stream.conn != nil {
+		s.grpcPool.ReleaseConn(stream.conn, r.storeAddr)
+		stream.conn = nil
+	}
+}
+
+func (s *SharedClient) runStream(
```

Comment:
It is recommended to move stream-related logic to a separate file.

Reply:
I prefer to keep it as a method of `SharedClient`.